rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -2031,6 +2037,330 @@ RMW_Connext_Subscriber::take_serialized(
+          ctx, node, this))
+      {
+        RMW_CONNEXT_LOG_ERROR("failed to update graph for subscriber")
+        return RMW_RET_ERROR;
+      }
+
+      this->status_condition->invalidate();
+
+      if (DDS_RETCODE_OK !=
+      DDS_Subscriber_delete_datareader(this->dds_subscriber(), this->dds_reader))
+      {
+        RMW_CONNEXT_LOG_ERROR_SET(
+          "failed to delete DDS DataReader")
+        return RMW_RET_ERROR;
+      }
+
+      // remove content filter topic
+      DDS_DomainParticipant * const dp = this->dds_participant();
+
+      if (nullptr != this->dds_topic_cft) {
+        rmw_ret_t cft_rc = rmw_connextdds_delete_contentfilteredtopic(
+          ctx, dp, this->dds_topic_cft);
+
+        if (RMW_RET_OK != cft_rc) {
+          return cft_rc;
+        }
+      }
+
+      this->dds_topic_cft = nullptr;
+
+      // create normal topic and a new reader
+      DDS_TopicDescription * sub_topic = DDS_Topic_as_topicdescription(dds_topic);
+
+      #if !RMW_CONNEXT_DDS_API_PRO_LEGACY
+        DDS_DataReaderQos dr_qos = DDS_DataReaderQos_INITIALIZER;
+      #else
+        DDS_DataReaderQos dr_qos;
+        if (DDS_RETCODE_OK != DDS_DataReaderQos_initialize(&dr_qos)) {
+          RMW_CONNEXT_LOG_ERROR_SET("failed to initialize datareader qos")
+          return RMW_RET_ERROR;
+        }
+      #endif /* !RMW_CONNEXT_DDS_API_PRO_LEGACY */
+
+      DDS_DataReaderQos * const dr_qos_ptr = &dr_qos;
+      auto scope_exit_dr_qos_delete =
+        rcpputils::make_scope_exit(
+        [dr_qos_ptr]()
+        {
+          DDS_DataReaderQos_finalize(dr_qos_ptr);
+        });
+
+      if (DDS_RETCODE_OK !=
+        DDS_Subscriber_get_default_datareader_qos(this->dds_subscriber(), &dr_qos))
+      {
+        RMW_CONNEXT_LOG_ERROR_SET("failed to get default reader QoS")
+        return RMW_RET_ERROR;
+      }
+
+      DDS_DataReader * dds_reader =
+        rmw_connextdds_create_datareader(
+        ctx,
+        dp,
+        this->dds_subscriber(),
+        &qos_policies,
+    #if RMW_CONNEXT_HAVE_OPTIONS_PUBSUB
+        nullptr,
+    #endif /* RMW_CONNEXT_HAVE_OPTIONS_PUBSUB */
+        internal,
+        type_support,
+        sub_topic,
+        &dr_qos);
+
+      if (nullptr == dds_reader) {
+        RMW_CONNEXT_LOG_ERROR_SET("failed to create DDS reader")
+        return RMW_RET_ERROR;
+      }
+
+      this->dds_reader = dds_reader;
+      this->status_condition.reset(new RMW_Connext_SubscriberStatusCondition(this->dds_reader, this->ignore_local));
+      rmw_connextdds_get_entity_gid(this->dds_reader, this->ros_gid);
+
+      if (!internal) {
+        if (RMW_RET_OK != this->enable()) {
+          RMW_CONNEXT_LOG_ERROR("failed to enable subscription")
+          return RMW_RET_ERROR;
+        }
+
+        if (RMW_RET_OK !=
+          rmw_connextdds_graph_on_subscriber_created(
+            ctx, node, this))
+        {
+          RMW_CONNEXT_LOG_ERROR("failed to update graph for subscriber")
+          return RMW_RET_ERROR;
```

Comment:
In this function, most of 'RMW_RET_ERROR' means the critical error.     
Maybe it better to add some descriptions on that user have better to destroy and re-create subscriber while getting return value 'RMW_RET_ERROR'.

Reply:
I used the same destroy source code from some destroy function to delete these resources. If it failed any place on this function, I don't think the user can use some correct function to destroy this subscriber.

Reply:
Get it.